Update nmethod caching and clean up unused/duplicate methods

ashu-mehra opened the pull request:
This PR updates nmethod caching to be based on caching of CodeBlob. It also cleans up some dead code and duplicate methods.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - Committer) ⚠️ Review applies to [31e581d4](https://git.openjdk.org/leyden/pull/71/files/31e581d4455e2acb4a91747d916bf4ccbb44562d)
 * [Andrew Dinn](https://openjdk.org/census#adinn) (@adinn - Committer)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/leyden.git pull/71/head:pull/71` \
`$ git checkout pull/71`

Update a local copy of the PR: \
`$ git checkout pull/71` \
`$ git pull https://git.openjdk.org/leyden.git pull/71/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 71`

View PR using the GUI difftool: \
`$ git pr show -t 71`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/leyden/pull/71.diff">https://git.openjdk.org/leyden/pull/71.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/leyden/pull/71#issuecomment-2930943816)
</details>

vnkozlov commented on src/hotspot/share/code/aotCodeCache.hpp:
```diff
@@ -102,12 +101,12 @@ class AOTCodeEntry {
-  uint   _reloc_size;  // Max size of relocations per code section
   uint   _num_inlined_bytecodes;
 
+  uint   _blob_offset; // Start of archived blob in the cache
+  uint   _code_offset; // Start of code for an entry of type "Stub"
```
Why you need 2 different fields for offset?

ashu-mehra replied:
I did this only for the clarity - using `_blob_offset` is for storing Stubs doesn't sound correct, so I added `_code_offset` for the Stubs. If it is confusing, I can keep `_blob_offset` and remove the `_code_offset` field.

vnkozlov replied:
Or we can compromise: `_codeblob_offset, _codeblob_size`
Stubs are also using CodeBlob. I don't get why you treat stubs differently.

ashu-mehra replied:
StubGenerator Stubs used CodeBlob, but in AOT CodeCache we don't store their CodeBlobs as we do for adapters/runtime blobs. We instead only copy the assembly code for each stub.

shipilev commented:
I wonder if both https://github.com/openjdk/leyden/pull/75 and https://github.com/openjdk/leyden/pull/76 preclude properly testing this PR. Let's integrate those first too.

vnkozlov commented:
> I wonder if both #75 and #76 preclude properly testing this PR. Let's integrate those first too.

Yes, and merge from mainline with all AOT bugs fixes there. I will wait JDK 25 fork before merge.

vnkozlov commented:
And we need to fix bootstrap sequence: [JDK-8358690](https://bugs.openjdk.org/browse/JDK-8358690) and https://github.com/openjdk/leyden/pull/75

ashu-mehra commented:
@vnkozlov @adinn I have merged premain and the fix for JDK-8358330. Also addressed review comments to remove `AOTCodeEntry::_blob_offset`. GHA is passing now. So I guess its ready for another round of review.

vnkozlov commented:
@ashu-mehra  unfortunately I don't have time to look on this in details now. Will look on it later. On first glance it is fine.

vnkozlov reviewed (commented):
My testing passed and everything seems fine.
Did you compare outputs logs that we produce/use about the same number of AOT nmethods?
Did you compare performance?

vnkozlov commented:
May be look on `XX:+CITime` output.

ashu-mehra commented:
I tested the changes with spring-boot-getting-started and JavacBench. The logs indicate number of nmethods stored and loaded are more or less the same. Performance change is also negligible or within margin of error.
For spring-boot-getting-started:
```
Run,Old CDS + AOT,New CDS + AOT
1,323,326
2,324,324
3,323,323
4,326,329
5,324,324
6,323,326
7,328,332
8,320,324
9,323,323
10,324,325
Geomean,323.79,325.59
Stdev,1.99,2.73
```
For JavacBench:
```
Run,Old CDS + AOT,New CDS + AOT
1,537,544
2,534,548
3,530,538
4,546,540
5,537,527
6,535,529
7,540,544
8,540,545
9,538,548
10,544,540
Geomean,538.08,540.26
Stdev,4.46,6.91
```

Comparing AOT Code Cache stats generated by Xlog:init for spring-boot-getting-started...
premain:
```
[0.316s][info][init] AOT Code Cache statistics (when closed): 
[0.316s][info][init]   Adapter: total=535
[0.316s][info][init]   Code: total=2826 loaded=2553 invalidated=194 failed=9
[0.316s][info][init]     AOT Code T1 total=131 loaded=127
[0.316s][info][init]     AOT Code T2 total=1870 loaded=1857 invalidated=18 failed=9
[0.316s][info][init]     AOT Code T4 total=415 loaded=173 invalidated=5
[0.316s][info][init]     AOT Code T5 total=410 loaded=396 invalidated=171 has_clinit_barriers=302
```
and with this patch:
```
[0.316s][info][init] AOT Code Cache statistics (when closed): 
[0.316s][info][init]   Adapter: total=535
[0.316s][info][init]   Code: total=2849 loaded=2557 invalidated=182 failed=9
[0.316s][info][init]     AOT Code T1 total=131 loaded=127
[0.316s][info][init]     AOT Code T2 total=1876 loaded=1863 invalidated=18 failed=9
[0.316s][info][init]     AOT Code T4 total=423 loaded=160 invalidated=6
[0.316s][info][init]     AOT Code T5 total=419 loaded=407 invalidated=158 has_clinit_barriers=311
```

Similarly for JavacBench, premain:
```
[0.270s][info][init] AOT Code Cache statistics (when closed): 
[0.270s][info][init]   Adapter: total=430
[0.270s][info][init]   Code: total=3049 loaded=2816 invalidated=322
[0.270s][info][init]     AOT Code T1 total=93 loaded=93
[0.270s][info][init]     AOT Code T2 total=1904 loaded=1899
[0.270s][info][init]     AOT Code T4 total=527 loaded=299 invalidated=24
[0.270s][info][init]     AOT Code T5 total=525 loaded=525 invalidated=298 has_clinit_barriers=359
```
and with this patch:
```
[0.246s][info][init] AOT Code Cache statistics (when closed): 
[0.246s][info][init]   Adapter: total=430
[0.246s][info][init]   Code: total=3077 loaded=2858 invalidated=385
[0.246s][info][init]     AOT Code T1 total=93 loaded=93
[0.246s][info][init]     AOT Code T2 total=1837 loaded=1832
[0.246s][info][init]     AOT Code T4 total=574 loaded=360 invalidated=25
[0.246s][info][init]     AOT Code T5 total=573 loaded=573 invalidated=360 has_clinit_barriers=409
```

vnkozlov reviewed (approved):
Good.

vnkozlov commented:
If @adinn don't have any more comments you can integrate.

We may need few more AOT fixes in mainline before marge from it.

ashu-mehra commented on src/hotspot/share/code/aotCodeCache.hpp:
```diff
@@ -120,51 +118,48 @@ class AOTCodeEntry {
+               uint comp_id = 0, uint decomp = 0,
+               bool has_clinit_barriers = false,
+               bool for_preload = false,
+               bool ignore_decompile = false) {
```
my bad, fixed it now.

vnkozlov commented:
@ashu-mehra  I submitted testing after this was integrated and one closed test crashed on MacOS-aarch64 during "product" run when GC trying to purge nmethod:

```
#  Internal Error (/workspace/open/src/hotspot/share/nmt/mallocHeader.inline.hpp:107), pid=55426, tid=16643
#  fatal error: NMT corruption: Block at 0x00000001209f4088: header canary broken
#
# JRE version: Java(TM) SE Runtime Environment (25.0) (fastdebug build 25-internal-LTS-2025-06-16-1707161.vladimir.kozlov.leyden)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 25-internal-LTS-2025-06-16-1707161.vladimir.kozlov.leyden, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)

Stack: [0x000000016e2fc000,0x000000016e4ff000],  sp=0x000000016e4fddf0,  free space=2055k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.dylib+0x127c010]  VMError::report(outputStream*, bool)+0x1b00  (mallocHeader.inline.hpp:107)
V  [libjvm.dylib+0x127f8b8]  VMError::report_and_die(int, char const*, char const*, char*, Thread*, unsigned char*, void const*, void const*, char const*, int, unsigned long)+0x55c
V  [libjvm.dylib+0x5cf650]  DebuggingContext::~DebuggingContext()+0x0
V  [libjvm.dylib+0xdeed80]  MallocHeader* MallocHeader::resolve_checked_impl<void*, MallocHeader*>(void*)+0x19c
V  [libjvm.dylib+0xdee630]  MallocTracker::record_free_block(void*)+0x44
V  [libjvm.dylib+0xf20f84]  os::free(void*)+0xa8
V  [libjvm.dylib+0xec194c]  nmethod::purge(bool)+0x158
V  [libjvm.dylib+0x4d5fe0]  ClassUnloadingContext::purge_nmethods()+0xac
V  [libjvm.dylib+0x71961c]  G1CollectedHeap::unload_classes_and_code(char const*, BoolObjectClosure*, GCTimer*)+0x434
V  [libjvm.dylib+0x72dd40]  G1ConcurrentMark::remark()+0x2b0
V  [libjvm.dylib+0x7f67e4]  VM_G1PauseConcurrent::doit()+0x1b8
V  [libjvm.dylib+0x128743c]  VM_Operation::evaluate()+0x11c
V  [libjvm.dylib+0x12a4e20]  VMThread::evaluate_operation(VM_Operation*)+0x114
V  [libjvm.dylib+0x12a5924]  VMThread::inner_execute(VM_Operation*)+0x320
V  [libjvm.dylib+0x12a4ad0]  VMThread::loop()+0x98
V  [libjvm.dylib+0x12a4868]  VMThread::run()+0xc0
V  [libjvm.dylib+0x11be0d4]  Thread::call_run()+0xf0
V  [libjvm.dylib+0xf28da4]  thread_native_entry(Thread*)+0x140
C  [libsystem_pthread.dylib+0x726c]  _pthread_start+0x94
VM_Operation (0x000000016dac2bf0): G1PauseRemark, mode: safepoint, requested by thread 0x0000000147841810
```

vnkozlov commented:
An other crash happened in aotCode/AOTCodeFlags.java on linux-amd64 in `find_entry()` which looks like during end of test run:
```
Event: 0.085 Thread 0x00007fc79002f0d0 Before exit entered
```

```
#  SIGSEGV (0xb) at pc=0x00007fc79680ead8, pid=3219122, tid=3219333
#
# JRE version: Java(TM) SE Runtime Environment (25.0) (fastdebug build 25-internal-LTS-2025-06-16-1707161.vladimir.kozlov.leyden)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 25-internal-LTS-2025-06-16-1707161.vladimir.kozlov.leyden, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x69cad8]  AOTCodeCache::find_entry(AOTCodeEntry::Kind, unsigned int, unsigned int, unsigned int)+0x158

Command Line: -XX:AOTCache=hello.aot -Xlog:aot -Xlog:aot+codecache* Hello

Host: ol8-x64-742696, AMD EPYC 7J13 64-Core Processor, 12 cores, 23G, Oracle Linux Server release 8.10
Time: Mon Jun 16 17:40:14 2025 UTC elapsed time: 0.094271 seconds (0d 0h 0m 0s)

---------------  T H R E A D  ---------------

Current thread (0x00007fc790249570):  JavaThread "Training replay thread" daemon [_thread_in_vm, id=3219333, stack(0x00007fc7701e6000,0x00007fc7702e6000) (1024K)]

Stack: [0x00007fc7701e6000,0x00007fc7702e6000],  sp=0x00007fc7702e3440,  free space=1013k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x69cad8]  AOTCodeCache::find_entry(AOTCodeEntry::Kind, unsigned int, unsigned int, unsigned int)+0x158  (aotCodeCache.cpp:1001)
V  [libjvm.so+0x69e09f]  AOTCodeCache::find_code_entry(methodHandle const&, unsigned int)+0x22f  (aotCodeCache.cpp:955)
V  [libjvm.so+0xb8cfc3]  CompileBroker::compile_method_base(methodHandle const&, int, int, int, CompileTask::CompileReason, bool, bool, Thread*)+0x103  (compileBroker.cpp:1428)
V  [libjvm.so+0xb8dd68]  CompileBroker::compile_method(methodHandle const&, int, int, int, bool, CompileTask::CompileReason, DirectiveSet*, JavaThread*)+0x4b8  (compileBroker.cpp:1729)
V  [libjvm.so+0xb8e40c]  CompileBroker::compile_method(methodHandle const&, int, int, int, bool, CompileTask::CompileReason, JavaThread*)+0xbc  (compileBroker.cpp:1614)
V  [libjvm.so+0xb4d29c]  CompilationPolicy::maybe_compile_early(methodHandle const&, JavaThread*) [clone .part.0]+0x19c  (compilationPolicy.cpp:120)
V  [libjvm.so+0xb4d78d]  CompilationPolicy::replay_training_at_init_impl(InstanceKlass*, JavaThread*)+0x3ad  (compilationPolicy.cpp:176)
V  [libjvm.so+0xb5244a]  CompilationPolicy::replay_training_at_init_loop(JavaThread*)+0x24a  (compilationPolicy.cpp:214)
V  [libjvm.so+0x110180b]  JavaThread::thread_main_inner()+0x13b  (javaThread.cpp:774)
V  [libjvm.so+0x1b946d6]  Thread::call_run()+0xb6  (thread.cpp:254)
V  [libjvm.so+0x180de89]  thread_native_entry(Thread*)+0x1a9  (os_linux.cpp:868)
```

vnkozlov commented:
Output before second failure (AOTCodeFlags.java):
```
[0.084s][info][aot,codecache,nmethod] 220 (L1): Reading nmethod 'java.lang.invoke.MethodType::returnType()Ljava/lang/Class;' (decomp: 0, hash: 0x49e35a1b, ignore_decomp)
[0.084s][info][aot,codecache        ] 220 (L1): Shared method lookup: java.lang.invoke.MethodType::returnType()Ljava/lang/Class;
[0.084s][info][aot,codecache,nmethod] 220 (L1): Read nmethod 'java.lang.invoke.MethodType::returnType()Ljava/lang/Class;' from AOT Code Cache
Hello World
#
# A fatal error has been detected by the Java Runtime Environment:
```

vnkozlov commented:
In mainline we don't do that.

vnkozlov commented:
Looks like in both case we closed AOT code cache but still trying to access it

vnkozlov commented:
> > I forgot that we call AOTCodeCache::close() on VM exit. It is old workflow. So we do explicitly closing it during production run. Should we even do that with new workflow?
> 
> If we can make AOTCodeCache accessible until exit, then I guess we can continue using immutable data and oop_map from the cache.

It seems work. We only loosing statistic output from AOTCodeClose::close(). We actually should not do it there. There is separate path for printing statistic `print_statistics()` on VM exit which will call AOT code statistics.

vnkozlov commented:
Hmm, are you suggesting to not do `delete _cache; _cache = nullptr;` only but still call `AOTCodeCache::close()` ?
This could work too.

vnkozlov commented:
I pushed fix: https://github.com/openjdk/leyden/commit/62a4bc7b77a311abca403e2be08aa16f5e2a80b0

vnkozlov commented:
> @vnkozlov
> 
> > I pushed fix: [62a4bc7](https://github.com/openjdk/leyden/commit/62a4bc7b77a311abca403e2be08aa16f5e2a80b0)
> 
> does this fix both the crashes that you observed?

I hope so. I did not see these failures anymore during new testing.
